fix/add-tests-for-url-shortner-project

abbathaw opened the pull request:
## This PR is closed after a git rebase mistake / disaster. New PR is here #39311 

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution. (I added attribution below).
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #37305

3 of 4 tests have been copied from this [location](https://glitch.com/edit/#!/fcc-back-end-tester?path=public/tests.js:76:236) owned by [christophrus](https://glitch.com/@christophrus) and I was directed to them as per discussions in #37185. 

I went through his tests and found that the example he provided to test the url body using the [fcc-backend-tester](https://glitch.com/~fcc-back-end-tester) was good enough to use and merge here.  

The 4th test is copied from a comment by @ojeytonwilliams in the same issue above.

raisedadead committed:
feat(learn): python certification projects (#38216)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Kristofer Koishigawa <scissorsneedfoodtoo@gmail.com>
Co-authored-by: Beau Carnes <beaucarnes@gmail.com>

moT01 commented on .gitignore:
```diff
@@ -3,6 +3,9 @@
 
+### Intellij IDEA ###
+/.idea/
+
```
I'm not familiar with this. I assume it won't be an issue, but a little explanation on why it's being added would have been nice.

abbathaw replied:
this is the settings file generated when using IntelliJ as a code editor. I will remove this.

moT01 commented on curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md:
```diff
@@ -11,6 +11,15 @@ forumTopicId: 301509
+      - It should return a json response with the shortUrl (e.g.`{"original_url":"www.google.com","short_url":1}`).
+1. Return json error message with message of `invalid url` if the url provided in the body of the above POST request isn't a valid URL. e.g. `res.json({"error":"No short url found for given input"})`
+1. Expose this GET endpoint: `/api/shorturl/{:shortUrl}` that will parse the shortUrl given in the params and redirect to the original Url.
+
```
```suggestion
```

I like adding instructions here, but they are listed in the project. I think if we want to add the user stories to these backend project challenges, we should make an issue to add them all at once so they are consistent.

abbathaw replied:
good point. The user stories are already in the readme file of the project that the learner will clone.

moT01 reviewed (changes requested):
Thanks for the PR @abbathaw.

The tests looks like they work, I'm not sure if we want to be attaching variables to the window object though. What do you think about that @ojeytonwilliams @RandellDawson?

RandellDawson commented:
@moT01 I do not like the window properties.  @abbathaw Can you not create local variables for the tests?

beaucarnes committed:
feat(learn): add python video challenges (#38424)

mstillman9 committed:
fix regex

moT01 committed:
feat: add video lessons to db on submit + update UI (#38591)

* feat: add video lessons to db on submit + update UI

* feat: delete CompletionVideoModal

* feat: clean up component + add comments

* feat: remove comment

* feat: remove log

* feat: remove log

* fix: update buttons + fix some testing

* fix: remove unused selector

beaucarnes committed:
add tensorflow course without questions (#38525)

beaucarnes committed:
feat: Add more video challenges (#38598)

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

moT01 committed:
fix(learn): tests for python video lessons (#38600)

* fix: improve tests for video challenges

* fix: remove unused challenges + get rid of no description warning

* fix: remove non-existent challenges from meta

moT01 committed:
fix: add video ids and better test for video id

moT01 committed:
fix: rename titles for better UX

moT01 committed:
fix: update missed title

moT01 committed:
fix:update video id + rename last file

moT01 committed:
fix: review and clean up a few things in this section

moT01 committed:
fix: misaligned titles (#38613)

beaucarnes committed:
feat(learn): split InforSec and QA into individual certs (#38347)

Co-authored-by: Mrugesh Mohapatra <me@raisedadead.com>
Co-authored-by: Kris Koishigawa <scissorsneedfoodtoo@gmail.com>

moT01 committed:
fix: rename tensorflow lessons (#38617)

moT01 committed:
fix: infosec show cert button not rendering (#38620)

raisedadead committed:
fix(client): nicer error message and layout update

moT01 committed:
feat: add loader on video challenges (#38633)

rayswebpresence committed:
Data Analysis Example A, Shape Property of DataFrames (#38654)

raisedadead committed:
fix: broken meta on challenges (#38681)

beaucarnes committed:
feat(learn): add Numpy video question. (#38619)

beaucarnes committed:
Feat: add python security video course (#38683)

niraliG committed:
feat(learn): add jupyter Notebook Cells video Question (#38628)

ojeytonwilliams committed:
feat: process video question md into html (#38667)

* feat: process video question md into html

* test: mdToHTML

* fix: use dedicated prism component

anastasiasoenjoto committed:
removed line breaks, fixed typos and wording of instructions

erictleung committed:
fix: Jupyter notebook challenge file name (#38740)

scissorsneedfoodtoo committed:
feat: add guides for chinese contributors

ojeytonwilliams committed:
fix(api): show most recent date for FullStack cert (#38736)

ShaunSHamilton committed:
fix(curriculum): change project codepen links (#38746)

levlaz committed:
docs: Update link to latest unanswered posts in the forum (#38742)

rayswebpresence committed:
Importing CSV Using Pandas Module (#38721)

* Importing CSV Using Pandas Module

* feat: process video question md into html (#38667)

* feat: process video question md into html

* test: mdToHTML

* fix: use dedicated prism component

* fix: Jupyter notebook challenge file name (#38740)

* fix quote error

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Eric Leung <eric@erictleung.com>
Co-authored-by: Beau Carnes <beaucarnes@gmail.com>

kylayujiri committed:
docs(docs): consolidate node.js instructions (#38749)

ojeytonwilliams committed:
test: record only if key is present (#38748)

raisedadead committed:
docs: re-arrange files for management

ShaunSHamilton committed:
fix(client): add relevant placeholder to link submission (#38411)

hassaanp committed:
fix (learn): add title to challenge meta for backend challenges (#38745)

UgRoss committed:
Fix: largestPrimeFactor function, Add: largestPrimeFactor(8) test (#38756)

* Fix `largestPrimeFactor(8)` result
* Improve `largestPrimeFactor` performance
* Add `largestPrimeFactor(8)` test

ERElli committed:
fix(learn): add test to return early pattern (#38751)

* add: test to return early pattern

Currently no test for a = 0 or b = 0. This results in if(a<=0 || b<= 0) also being a valid solution

* fix: typo in test text

* fix: update return early pattern test

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

anastasiasoenjoto committed:
fix(learn): Made Changes to Instructions in BFS/DFS exercise (#38500)

* changes to bfs instructions

* added description on visual representation

* Added to the animation description

* Added information about stacks in the description

* Edited description for clarity

* Edited small details for grammar and clarity

* Apply suggested wording and typo fixes

Co-authored-by: Eric Leung <eric@erictleung.com>

* Add typo fixes to BFS challenge

Co-authored-by: Kyla Yujiri <27860923+kylayujiri@users.noreply.github.com>
Co-authored-by: Eric Leung <eric@erictleung.com>

MarceloDamian committed:
fix(cypress): revised kill signals to processes in e2e.sh (#38692)

scissorsneedfoodtoo committed:
fix: dotall flag not supported in firefox (#38762)

Addresses https://github.com/freeCodeCamp/freeCodeCamp/pull/38546#issuecomment-625611675 where it's pointed out that Firefox doesn't support the /s regex flag yet. This workaround matches anything that's not blank.

ojeytonwilliams committed:
Feat(curriculum): scss not sass (#38764)

rayswebpresence committed:
Data Cleaning Duplicates: Duplicated Method (#38719)

rayswebpresence committed:
Read Data Fr Databases: Methods Available to a Cursor (#38720)

MarceloDamian committed:
fix(DOCS): Typo in comments (#38778)

raisedadead committed:
fix: 404 page styles (#38775)

raisedadead committed:
chore(api): bump minor & patch versions of deps

raisedadead committed:
chore(root): bump minor & patch versions of deps

raisedadead committed:
chore(client): bump minor & patch versions of deps

daksh-goyal committed:
Fixes a Typo in Project Euler Problem 70 (#38782)

* Fixed a Typo in Eulers Problem 70

Line 15 should say "Find the value of <var>n</var>, 1 < <var>n</var> < 10**7, for which φ(<var>n</var>) is a permutation of <var>n</var> and the ratio <var>n</var>/φ(<var>n</var>) produces a minimum." instead of Find the value of <var>n</var>, 1 < <var>n</var> < 107, for which φ(<var>n</var>) is a permutation of <var>n</var> and the ratio <var>n</var>/φ(<var>n</var>) produces a minimum.

* Updated the Markdown 

Fixes the markdown to display proper superscript rather than just using ** to denote a superscript

anastasiasoenjoto committed:
feat(learn): add fizzbuzz - RosettaCode challenges (#38610)

* create fizzbuzz challenge

* add lesson to meta file

* update testString

* fixed changes

vincentendrahadi committed:
fix(learn): improve landing page (#38695)

MarceloDamian committed:
docs: fixed header to match /learn (#38789)

moT01 committed:
feat: add markdown explanation to video challenge docs (#38758)

rayswebpresence committed:
feat(learn): Pandas Extracting Tables from HTML (#38726)

rayswebpresence committed:
feat(learn): NumPy Algebra and Size: Size and Efficiency (#38725)

* NumPy Algebra and Size: Size and Efficiency Question

* Update curriculum/challenges/english/08-data-analysis-with-python/data-analysis-with-python-course/numpy-algebra-and-size.english.md

Co-authored-by: Eric Leung <eric@erictleung.com>

Co-authored-by: Eric Leung <eric@erictleung.com>

rayswebpresence committed:
feat(learn): Denoting Code Blocks in Python (#38723)

rayswebpresence committed:
feat(learn): Data Analysis Example B - Using the 'Loc' Method (#38656)

MarceloDamian committed:
docs: changed headers to match learn part 2 (#38793)

UgRoss committed:
fix(client): prevent PayPal script loads in development (#38785)

raisedadead committed:
docs: update guide faq

raisedadead committed:
docs: update verbiage on guide faq

raisedadead committed:
docs: update verbiage in faq (#38810)

rayswebpresence committed:
Data Cleaning/Viz- Subplot Method Arguments (#38655)

* Data Cleaning/Viz- Subplot Method Arguments

* Update curriculum/challenges/english/08-data-analysis-with-python/data-analysis-with-python-course/data-cleaning-and-visualizations.english.md

Co-authored-by: Eric Leung <eric@erictleung.com>

* Update curriculum/challenges/english/08-data-analysis-with-python/data-analysis-with-python-course/data-cleaning-and-visualizations.english.md

Co-authored-by: Eric Leung <eric@erictleung.com>

* fix yaml

Co-authored-by: Beau Carnes <beaucarnes@gmail.com>
Co-authored-by: Eric Leung <eric@erictleung.com>

rayswebpresence committed:
Dictionary Iteration: Printing Key, Value Pairs (#38722)

* Dictionary Iteration: Printing Key, Value Pairs

* Apply suggestions from code review

Co-authored-by: Eric Leung <eric@erictleung.com>

Co-authored-by: Beau Carnes <beaucarnes@gmail.com>
Co-authored-by: Eric Leung <eric@erictleung.com>

raisedadead committed:
revert(api): bump minor & patch versions of deps (#38816)

raisedadead committed:
docs: update notes abut auth page

raisedadead committed:
feat(tools): add/update cleanup scripts (#38822)

AshNaz87 committed:
fix(curriculum): use example in Modifying an Object Within a Nested Object challenge (#38468)

* fix: use example in modifying an object within a nested object exercise to explain how to do so

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/modify-an-object-nested-within-an-object.english.md

Co-Authored-By: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/modify-an-object-nested-within-an-object.english.md

Co-Authored-By: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/modify-an-object-nested-within-an-object.english.md

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/modify-an-object-nested-within-an-object.english.md

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>
Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

anthonyzeng99 committed:
fix(learn): Fixed test issue in HashTable exercise (#38441)

* fixed issue where test for removal from hash table was passing even if method implementation is incorrect

ShaunSHamilton committed:
fix(curriculum): added test to allow any order solution (#38771)

* added test to allow any order solution

* add removeJSComments for tests

TyMick committed:
fix(curriculum): "Set Class" description formatting

TyMick committed:
Improve formatting and clarity in "Subset" (#38814)

* Improve formatting and clarity in "Subset"

* Change method name to isSubsetOf

Co-authored-by: Eric Leung <eric@erictleung.com>

* Change method name to isSubsetOf

Co-authored-by: Eric Leung <eric@erictleung.com>

* Fix method name in tests

* Fix variable name aIsSubsetOfB

* Fix method name in solution

Co-authored-by: Eric Leung <eric@erictleung.com>

vsasvipul0605 committed:
fix(curriculum): update solution to match seed (#38828)

dostrelith678 committed:
Allow use of g flag in Regular Expressions: Check for All or None (#38659)

* allow use of g flag

added functionality of tests to allow for the use of g flag by resetting the lastIndex as per example in https://github.com/freeCodeCamp/freeCodeCamp/pull/37941.

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/check-for-all-or-none.english.md

Co-Authored-By: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/check-for-all-or-none.english.md

Co-Authored-By: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/check-for-all-or-none.english.md

Co-Authored-By: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/check-for-all-or-none.english.md

Co-Authored-By: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

raisedadead committed:
fix(client): add footer on learn home (#38830)

luqven committed:
feat: show extensions warning in css learning module index (#38806)

raisedadead committed:
docs: update i18n faq (#38848)

paulywill committed:
Fix/curriculum en es6 complete a promise - modified test messages to be more clearer (#38836)

* Changed tests messages to indicate string

* minor tweak with wording of message

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

rena1234 committed:
fix: update regex specify exact number of matches - lastIndex = 0 before test (#38717)

ojeytonwilliams committed:
fix(docs): update i18n wording (#38849)

moT01 committed:
feat: add command for e2e tests to docs (#38851)

raisedadead committed:
fix(api): use mongodb+srv url protocol (#38831)

Kohota107 committed:
fix:(curriculum): Moved code tag to include new keyword (#38780)

* Moved code tag to include new keyword

The new keyword hasn't been explained in the curriculum yet, so have moved the <code> tag to include the new keyword to give a prompt that it needs to be used in the solution.

* Update curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-arrow-functions-to-write-concise-anonymous-functions.english.md

Added comma.

Co-authored-by: Lasse Jørgensen <28780271+lasjorg@users.noreply.github.com>

TyMick committed:
fix(curriculum): Clarify definition of "symmetric difference" (#38792)

* Clarify definition of "symmetric difference"

* Add definition of binary operation

* Adjust wording in "Symmetric Difference"

* Improve preposition use

* Move instructions to the Instructions section

Co-authored-by: Eric Leung <eric@erictleung.com>

ojeytonwilliams committed:
refactor(api): remove unused url-utils

ojeytonwilliams committed:
refactor(api): remove unused map util

ojeytonwilliams committed:
refactor(curriculum): remove old seed code

ojeytonwilliams committed:
feat: hide isHidden challenges

ojeytonwilliams committed:
fix(client): DRY out challenge sourcing

Also handles file changes the same way as the original sourcing.

ojeytonwilliams committed:
feat: restore challenge validation

rajat641 committed:
fix:Added extra test cases for 'arguments-optional' problem (#38827)

Twaha-Rahman committed:
fix(learn): remove cta and add current challenge button (#38807)

ojeytonwilliams committed:
fix(curriculum): control visibility via env

ojeytonwilliams committed:
fix(curriculum): add isHidden: false to challenges

This includes certificates (where it does nothing), but does not
include any translations.

raisedadead committed:
fix(learn): update project views for python

ojeytonwilliams committed:
fix: hide isHidden challenges from hotkey nav

arfbacem committed:
Add readme arabic (#38875)

Twaha-Rahman committed:
chore: add ESLint extension in .gitpod.yml (#38879)

raisedadead committed:
fix: move around new CIP challenges

Move newly added rosetta code challenges

raisedadead committed:
chore: update challengeType 11 to hidden

raisedadead committed:
chore: update challengeType 10 to hidden

raisedadead committed:
chore: update python, IS and QA cert challenge 7 to hidden

raisedadead committed:
chore: update new IS, QA challenge 4 to hidden

lasjorg committed:
fix(editor): Add quickSuggestions false to editor options (#38899)

moT01 committed:
fix: allow ctrl+enter to submit video lessons (#38898)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

moT01 committed:
fix: backend projects not submitting (#38900)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams committed:
fix(curriculum): bump Babel - compat with node lts (#38905)

theprakashkumar committed:
Update declare-javascript-variables.english.md (#38884)

ojeytonwilliams committed:
fix(curriculum): validate straight after parsing (#38896)

TyMick committed:
Clarify ambiguous instructions in "Stack Class" (#38802)

TyMick committed:
fix(learn): typo in "Priority Queue" description (#38804)

moT01 committed:
fix(client): add margin to certs (#38921)

moT01 committed:
fix: make videos bigger + add styles (#38924)

scissorsneedfoodtoo committed:
fix(client): landing page superblocks (#38914)

scissorsneedfoodtoo committed:
feat(curriculum): add python multiple choice questions (#38890)

ojeytonwilliams committed:
fix(client): ensure validation works during watch (#38936)

ojeytonwilliams committed:
refactor(learn): use graphQL filter for Map (#38913)

ojeytonwilliams committed:
refactor(client): simply landing query (#38935)

scissorsneedfoodtoo committed:
fix: exercise tracker solution URL (#38854)

scissorsneedfoodtoo committed:
fix: personal library solution URL (#38910)

scissorsneedfoodtoo committed:
fix(curriculum): update anonymous message board url (#38937)

DaVicki committed:
fix(curriculum): update instructions (#38865)

* fix(learn): update instructions

edited text - change "returns a random number" to "returns a random whole number" to be more clear on the problem statement.

Address review comments : add the keyword "whole" to description, to make it consistent with challenge title and Instructions.

raisedadead committed:
docs: remove instructions about next branch (#38940)

ShaunSHamilton committed:
fix: remove extra word (#38949)

ShaunSHamilton committed:
fix(curriculum): python for everybody (#38953)

* fix: dictionaries and loops

* fix: dictionaries common applications

* fix: intermediate strings

* fix: introduction elements of python

* fix: networking using urllib in python

* fix: objects a sample class

* fix: python dictionaries

* fix: python lists

* fix: strings and lists

* fix: web services xml

* fix: refactor python questions to md

* fix: refactor questions to md

* fix: remove extra word

Removed unnecessary word is

* fix: add code blocks to answers

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

* fix: format python code

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

lasjorg committed:
fix(doc): Correct link for "making changes and syncing" (#38956)

Fixing localhost link.

danwaltmorgan committed:
Update visualize-data-with-a-heat-map.english.md (#38869)

I changed the second User Story #16 to User story #17

ExperimentMonty committed:
Update wording of what `multiplyAll` should do in "Nesting For Loops" (#38852)

TyMick committed:
Fix false negatives in bubble sort test (#38801)

niteshseram committed:
fix: Data Analysis with Python- Incorrect answer for Numpy Operations (#38972)

* fix: Data Analysis with Python- Incorrect answer for Numpy Operations

* Update curriculum/challenges/english/08-data-analysis-with-python/data-analysis-with-python-course/numpy-operations.english.md

Change option from 2 to 1

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

cdadityang committed:
fix(curriculum) remove unnecessary `*/` from description (#38969)

huyenltnguyen committed:
fix(api): update url in email template (#38970)

matthewbub committed:
fix(client): update report user verbiage and link (#38954)

moT01 committed:
fix(client): background-color in code blocks of questions (#38968)

scissorsneedfoodtoo committed:
fix(client): public profile qa and infosec certification links (#38948)

niteshseram committed:
fix: numpy accessing and changing elements

moT01 committed:
fix(client): remove auto redirect on report user page (#38974)

JapneetSingh5 committed:
Fixes a broken link to 'undefined is a keyword' (#38897)

Fixes a broken link to 'undefined is a keyword' by replacing the wrong link with the correct one.

ShaunSHamilton committed:
fix(curriculum): add clarification to question and escape backslash (#38951)

* fix: add clarification to question

* fix: adjust word placement

Adjust word placement in python for everybody curriculum

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

Co-authored-by: Tom <20648924+moT01@users.noreply.github.com>

ojeytonwilliams committed:
fix(docs): emphasize use of Markdown in challenges (#38986)

ojeytonwilliams committed:
fix(client): allow numbers as answers (#38992)

Handles yaml floats.

scissorsneedfoodtoo committed:
feat(learn): add full stack certificate md and handle errors

Co-authored-by: Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

jasnkwcz committed:
fix(learn): clarify formatting requirements for multiple ternary operators challenge (#38993)

* fix(learn): clarify formatting requirements for multiple ternary operators challenge

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

Vardhu2706 committed:
fix(curriculum): Fixed typo APi to API (#38998)

* Fixed a typo APi => API

Co-authored-by: Vardhu2706 <avsv.vardhaman@gmail.com>

raisedadead committed:
fix(certs): add year to certs introduced in 2020 (#38912)

raisedadead committed:
docs: add Oliver to the platform dev team 🎉 (#39006)

scissorsneedfoodtoo committed:
feat(docs): add chinese news author application (#38891)

Add doc for author applications to Chinese /news. Also renamed docs in _sidebar.md so only there's only the Chinese version of the title.

naomi-lgbt committed:
fix(curriculum): remove whitespace typo (#39009)

UnevenCoder committed:
Fixed formatting inconsistency in issue-template (#39016)

niteshseram committed:
fix: more conditional structures python (#38981)

scissorsneedfoodtoo committed:
fix(curriculum): standardize fill-in-the-blank format and typos (#39011)

This PR changes the format of earlier fill-in-the-blank questions/answers to the one used in later questions. It also fixes a few typos.

naomi-lgbt committed:
fix(learn): function must return boolean, not string. (#39010)

rajatgupta24 committed:
Add: forumTopicId for rosetta code (#38855)

raisedadead committed:
feat: update cypress version and config (#38892)

ojeytonwilliams committed:
fix(client): sync Redux store and DB (#39034)

majestic-owl448 committed:
chore(GitHub): template for reporting issues with articles (#39035)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>

majestic-owl448 committed:
chore(client): removing a forgotten console.log (#39055)

ojeytonwilliams committed:
docs: /s/first timers welcome/first timers only/g (#39059)

raisedadead committed:
fix: start-server-and-test to organise e2e tests (#39036)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams committed:
chore: refactor and simplify testing (#39050)

erictleung committed:
fix(learn): use more precise abbreviation for ReLU (#39065)

beaucarnes committed:
fix: update video link for Relationship Building challenge. (#39062)

beaucarnes committed:
fix: update Colab link and wording machine learning challenges (#39061)

Dman247 committed:
fix(curriculum): Fixed Spelling Error. 'lightning speed' (#39056)

Commit:
fix(curriculum): improve test robustness (#39004)

* Add: // Only change code...comments

* Allow user to modify bob

raisedadead committed:
feat(donate): allow mutiple donations with warning (#39066)

ShaunSHamilton committed:
feat(docs): add wsl setup to docs (#38997)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams committed:
fix(client): ensure dom ready before testing (#39073)

jQuery challenges can fail seemingly at random. These changes should
prevent the race condition between a user's $( document ).ready() and
test evalution.

rdm100 committed:
fix: updated and consolidated quotes (#39069)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

majestic-owl448 committed:
fix: removing duplicate quotes, author names, unwanted quotes (#39074)

majestic-owl448 committed:
docs(i18n): update sections and links to italian README.md (#39067)

scissorsneedfoodtoo committed:
feat(learn): rename certificates, cert slugs, and cert flags (#39026)

scissorsneedfoodtoo committed:
fix(learn): update pages for new qa and infosec projects (#39032)

ojeytonwilliams committed:
fix(api): csurf to SameSite 'strict', https only (#39077)

Lax and http are probably sufficient, but if the stricter versions work
there's no harm using them.

scissorsneedfoodtoo committed:
fix(learn): versioning for new certificate slugs (#39087)

ojeytonwilliams committed:
feat: add python syntax highlighting (#39079)

To err on the side of caution, this explicitly includes all languages
that were implicitly included before using the Babel plugin.

ojeytonwilliams committed:
fix: style for video answers (#39090)

Vertically centers the radio buttons.
Fixes answer backgrounds.
Corrects answer padding on night mode.

borntofrappe committed:
fix comparing-and-sorting-tuples answer (#39027)

* fix comparing-and-sorting-tuples answer

apply flag to sorted function

* remove commas from numpy array

Co-authored-by: Shaun Hamilton <51722130+Sky020@users.noreply.github.com>

QuincyLarson committed:
fix(client): Remove non-motivational quotes and add better congratulatory quotes (#39102)

raisedadead committed:
feat: create codeql for static analysis (#39106)

majestic-owl448 committed:
docs: update links from next-python-projects to master branch (#39109)

User in the forum wanting to help found 404 error trying to follow the links, I am switching them to master branch as it is where those projects can be found now.

raisedadead committed:
chore: remove unused quotes files (#39107)

ojeytonwilliams committed:
fix(api): csrf over http in development (#39114)

Commit:
Remove example code from challenge seed (#39083)

Co-authored-by: Parth Parth <34807532+thecodingaviator@users.noreply.github.com>

ojeytonwilliams committed:
feat(api): get challenges directly from /curriculum

ojeytonwilliams committed:
refactor: remove old curriculum map from api

ojeytonwilliams committed:
chore: remove challenge seeding

raisedadead committed:
chore: enable python curriculum (#39118)

ahmaxed committed:
fix: set radio button selection to null (#39120)

cdadityang committed:
fix(curriculum): Add default white bg to tree DS image for better view (#39086)

* fix(curriculum): Add bg to tree DS image for better view

gitwinst committed:
fix: correct Alice Walker quote (#39142)

QuincyLarson committed:
feat: GitHub Sponsors :tada: (#39144)

raisedadead committed:
feat(client): /s/react-youtube/plyr-react/g (#39137)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Commit:
fix(learn): change forum link in help modal (#39143)

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>

ojeytonwilliams committed:
chore: update forum links in NavLinks (#39141)

Commit:
chore: replace forum links (#39139)

PeterF521 committed:
docs: fixed some orthographic errors and improvements(#39157)

Co-authored-by: pedrof521 <pedrof521@gmail.com>

raisedadead committed:
docs: add Randell to the platform dev team 🎉 (#39160)

JovaniTarnowski committed:
fix(learn): Organized the Certifications on Settings to match the main page (#39146)

* Organized the Certifications on Settings to match the main page

Co-authored-by: Mrugesh Mohapatra <1884376+raisedadead@users.noreply.github.com>
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

huyenltnguyen committed:
fix(learn): capitalization and grammar errors (#39151)

raisedadead committed:
revert(client): /s/react-youtube/plyr-react/g (#39162)

This reverts commit df6a4f649d762a347add209c7323c0f9bcf0288c.
We found mulitple issues with the integration and will rework on this.

moT01 committed:
fix(client): hide youtube suggestions (#39164)

ojeytonwilliams committed:
fix: replace . with - in slugs (#39168)

yavidor committed:
Fixed Spelling bug #39191 (#39193)

PeterF521 committed:
challenges: description improvement (#39197)

Co-authored-by: PeterF521 <peterf521@gmail.com>

ojeytonwilliams committed:
feat(client): show line numbers on video questions (#39207)

JovaniTarnowski committed:
learn(fix) - fixed grammar because it was confuse to understand the sentence (#39188)

* fix(learn): confusing grammar in copy array items using slice

greggubarev committed:
Fixing minor Russian grammatical (#39212)

ojeytonwilliams committed:
fix: remove whitespace to test solution (#39210)

greggubarev committed:
Fixing Russian grammatical (#39211)

TyMick committed:
Improve formatting in Max Heap lessons (#38835)

* Improve formatting in max heap lessons

* Switch to HTML ordered lists

TyMick committed:
Standardize indent size in Data Structures challenge seeds (#38803)

* Standardize indent size in "Priority Queue" seed

* Standardize indent size in "Circular Queue" seed

* Standardize indent size

* Standardize indents "Search within a Linked List"

* Standardize indents "Remove Elements from a Linked List by Index"

c-logan committed:
commit 7/8 Rosetta tokenize (#39213)

zbianca committed:
Fix text inconsistency (#39203)

* Fix text inconsistency

To match the eight ECMAScript Language Types on the previous paragraph.

Co-authored-by: moT01 <20648924+moT01@users.noreply.github.com>

moshfeu committed:
docs: Update "Responsive Web Design" Github link (#39134)

Current link leads to a 404 page in the old "learn" repository

ojeytonwilliams committed:
fix(curriculum): include stack with test errors (#39209)

Ashikpaul committed:
Fixed some minor typos (#39228)

Ashikpaul committed:
Fixed some minor typos (#39226)

Ashikpaul committed:
Fixed a minor typo (#39227)

Ashikpaul committed:
Fixed some minor typos (#39225)

* Fixed some minor typos

* Adding few more changes

Adding changes suggested by @RandellDawson.

Thank you @RandellDawson

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

Himanshunitrr committed:
Update devops.md (#39237)

Fixed some typos.

KartikSoneji committed:
fix: Specify indent_size = 2 globally in .editorconfig. (#39236)

raisedadead committed:
docs: update flight manuals (#39240)

Updated and unified workflows for working with various application VMs

Commit:
fix(learn): clarify description for finders keepers (#39068)

* fix(learn): clarify description for finders keepers

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>
Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

moT01 committed:
Fix/make mongodb instructions accurate (#39214)

* fix: make mongodb instructions accurate

ojeytonwilliams committed:
fix(client): improve accuracy of offline warning (#39249)

The old message made it sound like there was a problem with the server,
even though no attempt was being made to contact it.

RandellDawson committed:
fix(learn): updated catphotoapp links (Arabic) (#39256)

RandellDawson committed:
fix(learn): updated catphotoapp links (Spanish) (#39255)

RandellDawson committed:
fix(learn): updated catphotoapp links (Portuguese) (#39254)

RandellDawson committed:
fix(learn): updated catphotoapp links (Russian) (#39253)

RandellDawson committed:
fix(learn): updated catphotoapp links (Chinese) (#39252)

RandellDawson committed:
fix(learn): updated catphotoapp links (#39215)

ojeytonwilliams committed:
refactor(api): clean up certification logic (#39196)

TyMick committed:
Disambiguate "Priority Queue" instructions (#38805)

* Disambiguate "Priority Queue" instructions

* Add front() to solution

* Add tests for front() method

* Fix test typo

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

TyMick committed:
Add missing solutions for algorithm challenges (#38858)

TyMick committed:
Add <dfn> tags (#39260)

beaucarnes committed:
feat: add resources for data analysis videos (#39241)

* feat: add resources for data analysis videos

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

nakul251197 committed:
Fixed Some Typos (#39275)

ojeytonwilliams committed:
fix: responsive videos (#39171)

ojeytonwilliams committed:
refactor: test output -> arrays (#39178)

ojeytonwilliams committed:
chore: bump loadash and remove gulp (#39276)

Himanshunitrr committed:
Fixed-pass-an-array-as-props (#39268)

* Fixed-pass-an-arrays-props

* Fixed:pass-an-array-as-props2

* Update curriculum/challenges/english/03-front-end-libraries/react/pass-an-array-as-props.english.md

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

* Update curriculum/challenges/english/03-front-end-libraries/react/pass-an-array-as-props.english.md

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

* Update curriculum/challenges/english/03-front-end-libraries/react/pass-an-array-as-props.english.md

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

Co-authored-by: Randell Dawson <5313213+RandellDawson@users.noreply.github.com>

KartikSoneji committed:
docs: update link to `challengeTypes.js` (#39283)

NitronR committed:
feat(interview-prep): Converting and Tranfering Rosetta challenges (#38754)

* feat(interview-prep): Converting and Tranfering Rosetta challenges

scissorsneedfoodtoo committed:
fix: update donor email address and fix spacing

ojeytonwilliams committed:
fix: prevent preview-tests race condition (#39286)

ruhci28 committed:
Fix/delayofflinewarning (#39291)

* delay offline warning

* Delayed the offline warning

* delayed offline warning

* revert unwanted changes

* correct spacing

* correct spacing second time

* use let instead of var

KartikSoneji committed:
feat(gitpod): Add env variables to .bashrc so they can be accessed globally (#39296)

* feat(gitpod): Add env variables to .profile so they can be accessed globally.

abbathaw committed:
fix/add-tests-for-url-shortner remove global window variables in test
